dadguide/database_manager.py

```diff
@@ -201,10 +219,79 @@ def _get_table_fields(self, table_name: str):
             raise DadguideTableNotFound(table_name)
         return fields, pk
 
-    def get_active_skill(self, active_skill_id: int):
+    def build_graph(self):
+        self.graph = nx.DiGraph()
+
+        ms = self._query_many("SELECT * FROM monsters", (), DictWithAttrAccess)
+        es = self._query_many("SELECT * FROM evolutions", (), DictWithAttrAccess)
+        aws = self._query_many("SELECT * FROM awakenings", (), DgAwakening)
+        lss = self._query_many("SELECT * FROM leader_skills", (), DgLeaderSkill, idx_key='leader_skill_id')
+        ass = self._query_many("SELECT * FROM active_skills", (), DgActiveSkill, idx_key='active_skill_id')
+        ss = self._query_many("SELECT * FROM series", (), DgSeries, idx_key='series_id')
+
+        for m in ms:
+            self.graph.add_node(m.monster_id, **m,
+                                    dgawakenings=[],
+                                    dgls=lss.get(m.leader_skill_id),
+                                    dgas=ass.get(m.active_skill_id),
+                                    dgseries=ss.get(m.series_id)
+                                )
+            if m.linked_monster_id:
+                self.graph.add_edge(m.monster_id, m.linked_monster_id, type='transformation')
+                self.graph.add_edge(m.linked_monster_id, m.monster_id, type='back_transformation')
+
+        for e in es:
+            self.graph.add_edge(e.from_id, e.to_id, type='evolution', **e)
+            self.graph.add_edge(e.to_id, e.from_id, type='back_evolution', **e)
+
+        for a in aws:
+            self.graph.nodes[a.monster_id]['dgawakenings'].append(a)
+
+    def get_evo_tree(self, monster_id):
+        ids = set()
+        to_check = {monster_id}
+        while to_check:
+            mid = to_check.pop()
+            if mid in ids:
+                continue
+            n = self.graph[mid]
+            to_check.update(get_edges(n, 'evolution'))
+            to_check.update(get_edges(n, 'back_evolution'))
+            ids.add(mid)
+        return ids
+
+    def get_transform_tree(self, monster_id):
+        ids = set()
+        to_check = {monster_id}
+        while to_check:
+            mid = to_check.pop()
+            if mid in ids:
+                continue
+            n = self.graph[mid]
+            to_check.update(get_edges(n, 'transformation'))
+            to_check.update(get_edges(n, 'back_transformation'))
+            ids.add(mid)
+        return ids
+
+    def get_alt_cards(self, monster_id):
+        ids = set()
+        to_check = {monster_id}
+        while to_check:
+            mid = to_check.pop()
+            if mid in ids:
+                continue
+            n = self.graph[mid]
+            to_check.update(get_edges(n, 'evolution'))
+            to_check.update(get_edges(n, 'transformation'))
+            to_check.update(get_edges(n, 'back_evolution'))
+            to_check.update(get_edges(n, 'back_transformation'))
```

Comment:
doesn't this screw up dark magician girl and winged dragon? e.g `^id equip DMG`

Reply:
Yes, but I'm going to cross that bridge when it it comes up.  This seems like the best right now, because we don't know what direction GH is gonna take with transforms that can be obtained.  I can change it to do what would be best for specifically Winged Dragon's tree though.